core/server/master/src/main/java/alluxio/master/file/meta/InvalidationSyncCache.java

```diff
@@ -0,0 +1,418 @@
+ * appropriate invalidation times updated. Validation times are updated on the path
+ * when {@link #notifySyncedPath} is called on the root sync path after a successful sync.
+ * An invalidation is received either when a client calls
+ * {@link alluxio.master.file.DefaultFileSystemMaster#invalidateSyncPath} to notify that a path
+ * needs synchronization, or when a file is updated by an external Alluxio cluster, and
+ * cross cluster sync is enabled.
+ *
+ * Checking if a path needs to be synchronized involves checking the appropriate validation
+ * and invalidation times at each path component up to the root.
+ */
+@ThreadSafe
+public class InvalidationSyncCache {
+  private static final Logger LOG = LoggerFactory.getLogger(InvalidationSyncCache.class);
+
+  public final Cache<String, SyncState> mItems;
+  private final Clock mClock;
+  // we keep the root as a separate value as we never want to evict it,
+  // and instead we just update it under a lock
+  private final SyncState mRoot = new SyncState(false);
+  private final Lock mRootLock = new ReentrantLock();
+
+  /**
+   * Creates a new instance of {@link InvalidationSyncCache}.
+   * @param clock the clock to use to compute sync times
+   */
+  public InvalidationSyncCache(Clock clock) {
+    this(clock, null);
+  }
+
+  @VisibleForTesting
+  InvalidationSyncCache(Clock clock,
+      @Nullable BiConsumer<String, SyncState> onRemoval) {
+    mClock = clock;
+    mItems = CacheBuilder.newBuilder()
+        .removalListener(
+            (removal) -> {
+              if (removal.wasEvicted() && removal.getKey() != null && removal.getValue() != null) {
+                if (onRemoval != null) {
+                  onRemoval.accept((String) removal.getKey(), (SyncState) removal.getValue());
+                }
+                onCacheEviction((String) removal.getKey(), (SyncState) removal.getValue());
+              }
+            })
+        .maximumSize(Configuration.getInt(
+            PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY))
+        .build();
+  }
+
+  /**
+   * Called when an element is evicted from the cache.
+   * @param path the path
+   * @param state the state
+   */
+  void onCacheEviction(String path, SyncState state) {
+    try {
+      // on eviction, we must mark our parent as needing a sync with our invalidation time
+      notifyInvalidationInternal(PathUtils.getParent(path), state.mInvalidationTime);
+    } catch (InvalidPathException e) {
+      throw new RuntimeException("Should not have an invalid path in the cache", e);
+    }
+  }
+
+  /**
+   * There is a most one writer at a time to a sync state.
+   * But multiple readers may be reading the values while they
+   * are being updated, thus the values are volatile.
+   */
+  static class SyncState {
+    volatile long mDirectChildInvalidation = 0;
+    volatile long mRecursiveChildInvalidation = 0;
+    volatile long mInvalidationTime = 0;
+    volatile long mValidationTime = 0;
```

Comment:
does `ValidationTime` just mean the TS of last sync?

Reply:
yes

Reply:
I think that's hard to immediately understand a "validation" is a "sync". Would it make more sense to give up the name symmetric and just call it `lastSyncTime` vs `invalidationTime`?